### PR TITLE
Implement WITH queries

### PR DESCRIPTION
**Pull request by julianmenzler**
`WITH` is already supported by the `sql-parser`. This PR introduces support for it within the Hyrise SQL Translator.

### CODE REVIEW

File: src/lib/sql/sql_translator.cpp [hunk 1]
```diff
@@ -266,6 +277,26 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_select_statement(cons
...
+  SQLTranslator with_translator{_use_mvcc, nullptr, _parameter_id_allocator, _with_descriptions};
+  auto lqp = with_translator._translate_select_statement(*desc.select);
+
+  // Save mappings: ColumnID -> ColumnName
+  std::unordered_map<ColumnID, std::string> column_names;
+  for (auto column_id = ColumnID{0}; column_id < lqp->column_expressions().size(); ++column_id) {
+    const auto identifiers =
+        with_translator._sql_identifier_resolver->get_expression_identifiers(lqp->column_expressions()[column_id]);
+    for (const auto& identifier : identifiers) {
+      column_names.insert_or_assign(column_id, identifier.column_name);
```
Comment (julianmenzler) [inline src/lib/sql/sql_translator.cpp]: This is something I want to discuss with you @mrzzzrm today.

Comment (mrzzzrm) [reply src/lib/sql/sql_translator.cpp]: Has to be tomorrow, sorry. Today I won't be in

File: src/lib/sql/sql_translator.cpp [hunk 2]
```diff
@@ -130,7 +130,8 @@ bool is_trivial_join_predicate(const AbstractExpression& expression, const Abstr
 namespace opossum {
 
 SQLTranslator::SQLTranslator(const UseMvcc use_mvcc)
-    : SQLTranslator(use_mvcc, nullptr, std::make_shared<ParameterIDAllocator>()) {}
+    : SQLTranslator(use_mvcc, nullptr, std::make_shared<ParameterIDAllocator>(),
+                    std::unordered_map<std::string, std::shared_ptr<LQPView>>()) {}
```
Comment (mrzzzrm) [inline src/lib/sql/sql_translator.cpp]: `{}` would do

File: src/lib/sql/sql_translator.cpp [hunk 3]
```diff
@@ -191,20 +194,28 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_statement(const hsql:
...
+  // 7. SELECT clause (incl. DISTINCT)
+  // 8. UNION clause
+  // 9. ORDER BY clause
+  // 10. LIMIT clause
 
   AssertInput(select.selectList, "SELECT list needs to exist");
   AssertInput(!select.selectList->empty(), "SELECT list needs to have entries");
   AssertInput(!select.unionSelect, "Set operations (UNION/INTERSECT/...) are not supported yet");
 
+  // Translate WITH clause
```
Comment (mrzzzrm) [inline src/lib/sql/sql_translator.cpp]: clause**s**, right?

File: src/lib/sql/sql_translator.cpp [hunk 4]
```diff
@@ -191,20 +194,28 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_statement(const hsql:
...
+  if (select.withDescriptions) {
+    for (const auto& desc : *select.withDescriptions) {
+      _translate_hsql_with_description(*desc);
```
Comment (mrzzzrm) [inline src/lib/sql/sql_translator.cpp]: for consistent naming, I guess that should be called "_translate_hsql_with_description".

Since "desc" is a hyrise variable, it should have an unabbreviated name

File: src/lib/sql/sql_translator.cpp [hunk 5]
```diff
@@ -266,6 +277,26 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_select_statement(cons
   return _current_lqp;
 }
 
+void SQLTranslator::_translate_hsql_with_description(hsql::WithDescription& desc) {
+  SQLTranslator with_translator{_use_mvcc, nullptr, _parameter_id_allocator, _with_descriptions};
```
Comment (mrzzzrm) [inline src/lib/sql/sql_translator.cpp]: did you test placeholders in WITH? `WITH a AS SELECT * FROM t WHERE a > ? SELECT ...`?

File: src/lib/sql/sql_translator.cpp [hunk 6]
```diff
@@ -266,6 +277,26 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_select_statement(cons
...
+  auto lqp = with_translator._translate_select_statement(*desc.select);
```
Comment (mrzzzrm) [inline src/lib/sql/sql_translator.cpp]: can be const auto

File: src/lib/sql/sql_translator.cpp (hunk 1 above)
Comment (mrzzzrm) [reply src/lib/sql/sql_translator.cpp]: So basically you're invalidating all but one name of the column, right? How did you handle this for views @janetzki ?

File: src/lib/sql/sql_translator.cpp [hunk 7]
```diff
@@ -466,7 +497,27 @@ SQLTranslator::TableSourceState SQLTranslator::_translate_table_origin(const hsq
 
   switch (hsql_table_ref.type) {
     case hsql::kTableName: {
-      if (StorageManager::get().has_table(hsql_table_ref.name)) {
+      // WITH descriptions or sub-queries are treated as though they were inline views or tables
+      // They mask existing tables or views with the same name.
+      if (_with_descriptions.count(hsql_table_ref.name)) {
+        const auto iter = _with_descriptions.find(hsql_table_ref.name);
```
Comment (mrzzzrm) [inline src/lib/sql/sql_translator.cpp]: count() searches, find() searches - a single find() should be sufficient. Also, names iterators I like better, e.g. "with_descriptions_iterator"

File: src/lib/sql/sql_translator.cpp [hunk 8]
```diff
@@ -466,7 +497,27 @@ SQLTranslator::TableSourceState SQLTranslator::_translate_table_origin(const hsq
...
+        AssertInput(iter != _with_descriptions.end(), "No such WITH query named '" + hsql_table_ref.name + "'");
```
Comment (mrzzzrm) [inline src/lib/sql/sql_translator.cpp]: How could that ever trigger if `_with_descriptions.count(hsql_table_ref.name) > 0`, as the if below ensures?

File: src/test/sql/sql_translator_test.cpp [hunk 9]
```diff
@@ -1996,4 +1996,91 @@ TEST_F(SQLTranslatorTest, CatchInputErrors) {
...
+              stored_table_node_int_int_int));
+  // clang-format on
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(SQLTranslatorTest, WithClauseSingleQueryAlias) {
+  const auto actual_lqp_alias = compile_query(
+      "WITH "
+      "with_query AS (SELECT a AS x FROM int_int_int) "
+      "SELECT * FROM with_query WHERE x > 123;");
```
Comment (mrzzzrm) [inline src/test/sql/sql_translator_test.cpp]: `SELECT * FROM with_query WHERE x > 123` would throw an exception, currently, right? I think this should be fixed.

File: src/test/sql/sql_translator_test.cpp [hunk 10]
```diff
@@ -1996,4 +1996,91 @@ TEST_F(SQLTranslatorTest, CatchInputErrors) {
...
+  const auto aliases = std::vector<std::string>({"x"});
+  const auto expressions = expression_vector(int_int_int_a);
+
+  // clang-format off
+  const auto expected_lqp_alias =
+      AliasNode::make(expressions, aliases,
```
Comment (mrzzzrm) [inline src/test/sql/sql_translator_test.cpp]: indent of 2 as we do everywhere else should be preferred.

File: src/test/sql/sql_translator_test.cpp [hunk 11]
```diff
@@ -1996,4 +1996,91 @@ TEST_F(SQLTranslatorTest, CatchInputErrors) {
...
+                      stored_table_node_int_int_int))));
+  // clang-format on
+  EXPECT_LQP_EQ(actual_lqp_alias, expected_lqp_alias);
+}
+
+TEST_F(SQLTranslatorTest, WithClauseDoubleQuery) {
+  // WITH queries build on top of each other
+  const auto actual_lqp = compile_query(
+      "WITH "
+      "with_query1 AS (SELECT a, b FROM int_int_int), "
```
Comment (mrzzzrm) [inline src/test/sql/sql_translator_test.cpp]: a WHERE here and a WHERE in the line below would make this test a bit more interesting.

File: src/test/sql/sql_translator_test.cpp [hunk 12]
```diff
@@ -1996,4 +1996,91 @@ TEST_F(SQLTranslatorTest, CatchInputErrors) {
...
+  // clang-format on
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
+TEST_F(SQLTranslatorTest, WithClauseDoubleQueryMasking) {
+  // Second WITH query masks first one
+  const auto actual_lqp = compile_query(
+      "WITH "
+      "with_query AS (SELECT a, b FROM int_int_int), "
+      "with_query AS (SELECT b FROM with_query) "
```
Comment (mrzzzrm) [inline src/test/sql/sql_translator_test.cpp]: do other dbs allow this? if not, we shouldn't either, probably.

File: src/test/sql/sql_translator_test.cpp [hunk 13]
```diff
@@ -1996,4 +1996,91 @@ TEST_F(SQLTranslatorTest, CatchInputErrors) {
...
+
+  // clang-format off
+  const auto expected_lqp =
+      ProjectionNode::make(expression_vector(int_int_int_b),
+           ProjectionNode::make(expression_vector(int_int_int_a, int_int_int_b),
+               stored_table_node_int_int_int));
+  // clang-format on
+  EXPECT_LQP_EQ(actual_lqp, expected_lqp);
+}
+
```
Comment (mrzzzrm) [inline src/test/sql/sql_translator_test.cpp]: random line

why was submodule third_party/tpcds-result-reproduction  changed?

File: src/lib/sql/sql_translator.cpp (hunk 3 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: I thought of the WITH clause as a whole with multiple statements or descriptions: 
```SQL
WITH
a AS (..), 
b AS (..), 
c AS (..)
```
But this is my interpretation.

File: src/lib/sql/sql_translator.cpp (hunk 1 above)
Comment (janetzki) [reply src/lib/sql/sql_translator.cpp]: Good point. I did not make any changes in `_translate_create_view()` so far. I'll update you as soon as I have tested `CREATE VIEW` for hyrise-mp/hyrise#38.

File: src/lib/sql/sql_translator.cpp (hunk 3 above)
Comment (mrzzzrm) [reply src/lib/sql/sql_translator.cpp]: Ah, I was mistaken about the syntax. If there is ever only one WITH, leave it at "clause"

File: src/test/sql/sql_translator_test.cpp (hunk 12 above)
Comment (julianmenzler) [reply src/test/sql/sql_translator_test.cpp]: Unfortunately I do not have any offical SQL reference. But I had success testing this query with `Oracle11g` (via [sqltest.net](https://sqltest.net). 
But, I just noticed that the results Oracle11g returns are not correct or as expected. Oracle11g seems to use the first with_description, ignoring the second re-declaration.
I guess we should prohibits this kind of query then.

File: src/test/sql/sql_translator_test.cpp (hunk 13 above)
Comment (julianmenzler) [reply src/test/sql/sql_translator_test.cpp]: > why was submodule third_party/tpcds-result-reproduction changed?
No reason. Slipped in, somehow.

File: src/lib/sql/sql_translator.cpp (hunk 4 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: > for consistent naming, I guess that should be called "_translate_hsql_with_description".

Am I blind, I do not see any difference.

> Since "desc" is a hyrise variable, it should have an unabbreviated name

So `*with_description` instead of `*desc`?

File: src/lib/sql/sql_translator.cpp (hunk 5 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Oh these placeholders are for prepared statements, right? Not yet.

File: src/lib/sql/sql_translator.cpp (hunk 7 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Good catch!

File: src/lib/sql/sql_translator.cpp (hunk 8 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Correct 🙈

File: src/test/sql/sql_translator_test.cpp (hunk 9 above)
Comment (julianmenzler) [reply src/test/sql/sql_translator_test.cpp]: I do not quite understand.. Do you mean the `;` ?

File: src/lib/sql/sql_translator.cpp (hunk 4 above)
Comment (mrks) [reply src/lib/sql/sql_translator.cpp]: > I may be blind, but I do not see the difference.

Neither do I

> So *with_description instead of *desc?

Yes.

File: src/lib/sql/sql_translator.cpp (hunk 5 above)
Comment (mrks) [reply src/lib/sql/sql_translator.cpp]: ping?

File: src/lib/sql/sql_translator.cpp (hunk 6 above)
Comment (mrks) [reply src/lib/sql/sql_translator.cpp]: This was marked as resolved but doesn't look like it was.

File: src/lib/sql/sql_translator.cpp (hunk 7 above)
Comment (mrks) [reply src/lib/sql/sql_translator.cpp]: This PR looks like the comments have been seen, but not addressed yet. Skipping the rest of the comments and my own review for now.

File: src/lib/sql/sql_translator.cpp (hunk 1 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Did you update LQPView yet? @janetzki

File: src/lib/sql/sql_translator.cpp (hunk 6 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Now it is.

File: src/lib/sql/sql_translator.cpp (hunk 7 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Correct, I just started addressing them.

File: src/lib/sql/sql_translator.cpp (hunk 1 above)
Comment (janetzki) [reply src/lib/sql/sql_translator.cpp]: Yes, in https://github.com/hyrise/hyrise/pull/1692/commits/a926f691958e8cd2b2ffca648843d8275dee2eb5.

**Comment by mrks** [comment]
What's the status here?

File: src/lib/sql/sql_translator.cpp [hunk 14]
```diff
@@ -466,7 +497,26 @@ SQLTranslator::TableSourceState SQLTranslator::_translate_table_origin(const hsq
 
   switch (hsql_table_ref.type) {
     case hsql::kTableName: {
-      if (StorageManager::get().has_table(hsql_table_ref.name)) {
+
+      // WITH descriptions or sub-queries are treated as though they were inline views or tables
```
Comment (janetzki) [inline src/lib/sql/sql_translator.cpp]:
```suggestion
      // WITH descriptions or subqueries are treated as though they were inline views or tables.
```

File: src/lib/sql/sql_translator.cpp [hunk 15]
```diff
@@ -266,6 +277,26 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::_translate_select_statement(cons
...
 
+void SQLTranslator::_translate_hsql_with_description(hsql::WithDescription& desc) {
+  SQLTranslator with_translator{_use_mvcc, nullptr, _parameter_id_allocator, _with_descriptions};
+  const auto lqp = with_translator._translate_select_statement(*desc.select);
+
+  // Save mappings: ColumnID -> ColumnName
+  std::unordered_map<ColumnID, std::string> column_names;
+  for (auto column_id = ColumnID{0}; column_id < lqp->column_expressions().size(); ++column_id) {
+    const auto identifiers =
+        with_translator._sql_identifier_resolver->get_expression_identifiers(lqp->column_expressions()[column_id]);
```
Comment (janetzki) [inline src/lib/sql/sql_translator.cpp]: This empty line seems odd.

File: src/test/sql/sql_translator_test.cpp [hunk 16]
```diff
@@ -1996,4 +1996,91 @@ TEST_F(SQLTranslatorTest, CatchInputErrors) {
...
+TEST_F(SQLTranslatorTest, WithClauseSingleQuerySimple) {
+  const auto actual_lqp = compile_query(
+      "WITH "
+      "with_query AS (SELECT a, b FROM int_int_int) "
+      "SELECT * FROM with_query WHERE a > 123;");
+  // clang-format off
+  const auto expected_lqp =
+      PredicateNode::make(greater_than_(int_int_int_a, value_(123)),
+        ProjectionNode::make(expression_vector(int_int_int_a, int_int_int_b),
+          stored_table_node_int_int_int));
```
Comment (janetzki) [inline src/test/sql/sql_translator_test.cpp]: Please fix the indentation.

File: src/lib/sql/sql_translator.cpp (hunk 15 above)
Comment (janetzki) [reply src/lib/sql/sql_translator.cpp]: You might want to use `_inflated_select_list_elements[column_id].identifiers` instead of `get_expression_identifiers(...)` since `get_expression_identifiers(...)` would return `{"a", "a1", "a2"}` for both `ColumnID`s in `SELECT a AS a1, a AS a2`.

File: src/lib/sql/sql_translator.cpp (hunk 5 above)
Comment (mrks) [reply src/lib/sql/sql_translator.cpp]: ping ping

File: src/test/sql/sql_translator_test.cpp (hunk 9 above)
Comment (mrks) [reply src/test/sql/sql_translator_test.cpp]: Can't reproduce what was meant either. Maybe he was thinking about `a > 123`, but that is addressed in a different issue.

File: src/lib/sql/sql_translator.cpp (hunk 5 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Currently, I'm try to understand the placeholder implementation to create a test afterwards.

File: src/test/sql/sql_translator_test.cpp (hunk 9 above)
Comment (mrzzzrm) [reply src/test/sql/sql_translator_test.cpp]: He probably meant `a >123` ;)

File: src/lib/sql/sql_translator.cpp (hunk 15 above)
Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: How do you obtain identifiers from `_inflated_select_list_expressions`?
```
std::vector<std::shared_ptr<AbstractExpression>> _inflated_select_list_expressions
``` 

Unfortunately, I don't get your idea.

Comment (julianmenzler) [reply src/lib/sql/sql_translator.cpp]: Good point. 
For this change I have to merge the current master though. Before, we should merge #1753 to avoid migration efforts regarding `LQPView`.

**Comment by julianmenzler** [comment]
Except for @janetzki's last suggestion, blocked by #1753, all change requests should be adressed.

File: src/lib/sql/sql_translator.cpp (hunk 15 above)
Comment (janetzki) [reply src/lib/sql/sql_translator.cpp]: I have to trust you that you will address this because I will not be available until Aug. 6. ;)

**Comment by mrks** [comment]
@julianmenzler Is this ready to go once the merge conflict is resolved?

**Comment by mrks** [comment]
Ah, never mind. That was just the master merge.

**Comment by julianmenzler** [comment]
Ready to go

**Comment by mrks** [comment]
It sounds as if a5632c4 fixed a bug for which @janetzki provided an example. Does it make sense to add a test that covers this?

**Comment by julianmenzler** [comment]
Correct. It is rather an edge case though. The test should be short, so I'll write one.
Another thought came to my mind regarding this. I have not tested aggregate functions yet.

**Comment by julianmenzler** [comment]
I just added two more tests for WITH queries containing aggregate functions. 

> It sounds as if [a5632c4](https://github.com/hyrise/hyrise/commit/a5632c41883687efafce0fe819d747bb2b371d6e) fixed a bug for which @janetzki provided an example. Does it make sense to add a test that covers this?

I created a test and found a bug in the `SQLTranslator::_translate_table_origin`. The bug affects regular views as well (#1768). It does not make sense to address the bug in this PR. I will fix it separately and add the associated WITH test afterwards.
